### ISSUE TITLE
plugin: is it possible to recover user/bank information on a plugin reload?

### ISSUE THREAD

**Issue by cmoussa1** [flux-framework/flux-accounting#221]
In an offline conversation with @ryanday36, he mentioned that when the priority plugin is re-loaded after a `systemctl restart flux`, `flux account-priority-update` had to be run again in order to re-populate the plugin with the user/bank information from the flux-accounting database. A suggestion was made to see if the data sent to the plugin could survive a restart so that the `flux account-priority-update` command does not have to be run again.

A couple of thoughts came to mind as I was thinking about this:

1) Is it possible to run flux commands in conjunction with loading a plugin? i.e, is it possible to automatically call `flux account-priority-update` after the plugin is loaded from a TOML file?

2) Is there a way to save the state of the plugin before Flux restarts? In this case, I don't think there would be a requirement to resend database information since it would already have user/bank information before the instance restarted.

3) When a `cron` job is set up, does the command set to be run by `cron` run instantly? I know our [Admin Guide](https://flux-framework.readthedocs.io/en/latest/adminguide.html) lists the set of flux-accounting commands to be run periodically through `cron`, so maybe I need to add a note in the guide saying that these flux-accounting commands should be set up just after loading the multi-factor priority plugin?

Regardless, maybe I should add some clearer documentation stating that the plugin won't allow jobs to run if it doesn't have any user/bank information from the database.

**Comment by grondo** [comment]
It occurs to me a nice side-effect of storing some state in the KVS is that sysadmins and @ryanday36 could easily peek at the current JSON object and perhaps get some idea of the current state of the priority plugin.

**Comment by cmoussa1** [comment]
Ah, nice, thanks @grondo for taking the time to answer these thoughts! I wasn't really sure what approach to take so I was just thinking of random ideas here, and your response to each helped provide a lot of insight.

> Another potential idea here would be to switch `flux account-priority-update` to use the KVS instead of a direct RPC to the priority plugin. The command would write or update a JSON object in the KVS at a well known location. The priority plugin could then use a KVS watch instead of an RPC to grab the updated information.
>
> The benefit is that the latest data is always in the KVS, and the priority plugin only needs one method to read the user/bank information. A drawback is that, unless you did something clever, the entire set of user/bank information would need to be written to the KVS on each update, which might be less efficient that what you are doing now (if you are only sending changes in information as updates).

This sounds intriguing and definitely worth the work. Is there an example somewhere in flux-core of writing JSON objects to the KVS? The `flux account-priority-update` command currently creates a JSON object in Python and sends an RPC, so maybe there it can instead write a similar object in the KVS, like you mentioned. Where's a typical location to write these objects? Is `/var/lib/flux/` considered a well known location? I believe that's where the flux-accounting DB and job-archive DB are stored.

**Comment by grondo** [comment]
The KVS is a separate service from the sqlite databases stored in `/var/lib/flux`. It offers a way to associate values with string keys within the instance, see the [`flux-kvs(1)`](https://flux-framework.readthedocs.io/projects/flux-core/en/latest/man1/flux-kvs.html) man page for a high level overview if you are not familiar with it.

If you use `flux kvs dir` on an instance, you'll see that there are a few high level "directories" (the kvs uses a dotted path notation for directories, e.g. `resource.R` is the key `"R"` in the directory `resource`.) It would makes sense to put accounting information in an `accounting.` top-level directory at first.

What is the form of the JSON object currently sent to the plugin by `flux account-priority-update`? If you write to the same key in the KVS each time, e.g. `accounting.last-update` or something, note that the most recent JSON object will overwrite the previous one, so unless the JSON object contains _all_ the information the plugin needs, then it might not be a good strategy to do it this way.

Also, it occurs to me that if the object is overwritten each time, then that could result in a lot of unused content blobs over time, filling up disk space in the content store. This isn't as big of a problem now that we have a strategy for garbage collection, but perhaps @garlick should comment here.

A common way we've saved state in other services is by using an eventlog. In this case, instead of overwriting a KVS object each time with the entire new state, you would create an eventlog entry with only the changed state (not sure how feasible this is for the accounting plugin). The plugin would then subscribe to eventlog updates and during normal operation would just process the most recent event. On a restart it would process all events and therefore get from a clean state to the last state. I guess the drawback with this approach is that, over time, it will take longer and longer to initialize from the eventlog (though in practice I'm not sure if that would be a problem or not). The benefit is that you'd be using a well-practiced pattern, the eventlog data might be interesting in and of itself (bank information over time), and `flux account-priority-update` would only need to write updates instead of the entire state on each run.

**Comment by cmoussa1** [comment]
> What is the form of the JSON object currently sent to the plugin by `flux account-priority-update`?

The current object sent to the plugin is one object called `data` that just has a list of user/bank rows containing the relevant information needed to calculate job priorities and enforce limits, e.g:

```json
{
  "data" : [
    { "userid" : 1, "bank" : "b", "def_bank" : "def", "fairshare" : 0.5, "max_running_jobs" : 1, "max_active_jobs" : 5 },
    { "userid" : 2, "bank" : "b", "def_bank" : "def", "fairshare" : 0.5, "max_running_jobs" : 1, "max_active_jobs" : 5 },
  ]
}
```

I think more data (like basic `queue` information or `plugin_factor_weight` information) will become a part of `flux account-priority-update` in the future, so maybe it could be added to this same object; maybe something like:

```json
{
  "data" : [
    { "userid" : 1, "bank" : "b", "def_bank" : "def", "fairshare" : 0.5, "max_running_jobs" : 1, "max_active_jobs" : 5 },
    { "userid" : 2, "bank" : "b", "def_bank" : "def", "fairshare" : 0.5, "max_running_jobs" : 1, "max_active_jobs" : 5 },
  ],
  "queue_data": [
    { "queue": "queue1", "min_nodes_per_job": 1, "max_nodes_per_job": 1, "max_time_per_job": 1, "priority": 100 }
  ],
  "plugin_factor_data": [
    { "factor": "factor", "weight": 1 }
  ]
}
```

**Comment by grondo** [comment]
I guess the question is whether the entire set of data is sent in the update each time? I.e. does the object contain information for every known user/bank combination?

**Comment by garlick** [comment]
I kind of liked the idea of trying to get the python script kicked off right away when the instance starts up.  In addition to not requiring any new code, it would handle these cases well:
- the first time flux is started after configuring acct, when there is no acct "checkpoint" to restore (I assume now there is a manual step, or jobs sit in PRIORITY state until the script finally runs on its own)
- after a long maintenance window, where checkpoint could be far out of date, and jobs might start running based on the wrong data

If it can be done just by changing the cron config, that would be cool.

Nothing wrong with a plugin saving/restoring state to the kvs.  It just seems like the latest, greatest data is already in stable storage in the acct db so it might not be the most straightforward way to go about it.

**Comment by grondo** [comment]
Those are good points @garlick. Now that you mention it, it does seem a lot more straightforward to (somehow) coerce the update script to run right away. Sorry for leading you astray @cmoussa1!

I don't think the crontab format supports a way to run something now, then in the future when the given crontab entry matches. What about just an rc1 script as proposed above?

**Comment by cmoussa1** [comment]
Thanks @garlick for giving your thoughts, and that's completely okay @grondo! It was worth it for me to talk those ideas out, so thank you both for your input. An rc1 script seems like a good idea to me.

**Comment by grondo** [comment]
Note that the `rc1` script will only run when the instance is started or restarted. If the plugin is manually reloaded there would need to be instructions to rerun the update by hand or something.

**Comment by grondo** [comment]
BTW, just to be sure, the priority plugin holds (or sets priority unavailable) all jobs for which it doesn't have associated user/bank info right? And then once the bank information is updated it iterates through all held jobs to see if they can be released from PRIORITY state? If so, then we don't have to worry about races between the plugin being loaded and called on existing pending jobs and when the user or rc1 gets a chance to run the update script.

**Comment by cmoussa1** [comment]
> Note that the `rc1` script will only run when the instance is started or restarted. If the plugin is manually reloaded there would need to be instructions to rerun the update by hand or something.

Thanks for this note. I should add a note or something in the Admin Guide if it does not exist already.

> BTW, just to be sure, the priority plugin holds (or sets priority unavailable) all jobs for which it doesn't have associated user/bank info right?

The current behavior of the plugin is to reject submitted jobs who don't have an associated user/bank entry in the plugin.

**Comment by grondo** [comment]
> The current behavior of the plugin is to reject submitted jobs who don't have an associated user/bank entry in the plugin.

We might want to rethink this approach. I think the design was for jobs for which a priority cannot currently be calculated to be put in the PRIORITY state via a priority plugin returning `flux_jobtap_priority_unavail()` in the `job.state.priority` [callback topic](https://flux-framework.readthedocs.io/projects/flux-core/en/latest/man7/flux-jobtap-plugins.html#callback-topics).

When the priority plugin gets its first update, then it can try to reprioritize all jobs with a call to `flux_jobtap_reprioritize_all()`, which will in turn invoke the `job.priority.get` callback topic for all pending jobs.

However, I can't remember if/why we settled on rejecting jobs instead, so forgive me if I've forgotten an important detail. However, if jobs _do_ get rejected when there isn't user/bank information, then I think we have a synchronization problem between the job accounting db and priority plugin which we'll have to address (and I'm not sure how to do it)

**Comment by cmoussa1** [comment]
>However, I can't remember if/why we settled on rejecting jobs instead, so forgive me if I've forgotten an important detail. However, if jobs do get rejected when there isn't user/bank information, then I think we have a synchronization problem between the job accounting db and priority plugin which we'll have to address (and I'm not sure how to do it)

I believe the reason I gave for why I proposed rejecting jobs when a user can't be found in the plugin was something along the lines of: only those who have a user/bank entry in the flux-accounting database (i.e those who have a bank to charge jobs against) would be allowed to submit jobs. Since all of the user/bank combos in the `association_table` get sent to the plugin with `flux account-priority-update`, the plugin should receive all available data from the DB when the plugin is first loaded and the first update occurs. Those who do not exist in the database would, in my mind, not be allowed to submit jobs, since they don't have a bank to charge jobs against and/or get a job usage value that would affect their (and other users') fairshare value. 

However, I could definitely see the argument for instead returning `flux_jobtap_priority_unavail ()`; I think this would handle the case where a user/bank entry is added _after_ the initial `flux account-priority-update` but _before_ the first automatic update via `flux cron` to avoid a job rejection. But either way, a sys admin or scheduler operator would have to manually re-run the `flux account-priority-update` in the case where a new user/bank entry is added and that user needs to run jobs right away, right? Otherwise, would they have to wait for the next automatic update to get their jobs reprioritized via `flux_jobtap_reprioritize_all ()`? Maybe I have that wrong.

**Comment by grondo** [comment]
> But either way, a sys admin or scheduler operator would have to manually re-run the flux account-priority-update in the case where a new user/bank entry is added and that user needs to run jobs right away, right? Otherwise, would they have to wait for the next automatic update to get their jobs reprioritized via flux_jobtap_reprioritize_all ()? Maybe I have that wrong.

In either case, a user's jobs would not be able to run until their information is first added to the priority plugin. The difference here is whether the user's jobs are rejected outright or held until the information is updated.

On a system restart, the priority plugin is loaded, then somehow _all_ the user/bank information is updated lazily in the background (whether through `rc1` script or `cron` etc). Therefore, currently _all_ user submissions will be rejected until this first update happens. I think we're saying it would be better to hold all those new jobs in PRIORITY state until that first update, rather than having all the jobs be rejected.

**Comment by garlick** [comment]
Maybe the plugin could distinguish between 
- waiting for first accounting update (no user/bank data available yet)
- one or more updates has occurred, but job has no user/bank data

There's a strong argument for keeping jobs in PRIORITY in the first case since otherwise in the early moments after  system restart, all the pending jobs from before might get rejected.

**Comment by grondo** [comment]
> all the pending jobs from before might get rejected.

I don't think jobs that are already pending on a restart can get rejected since the `job.validate` callback is only called from the job-manager's `submit_cb()`. (It also wouldn't make sense because rejected jobs have to send a response back to the submitter, so it doesn't make sense to reject jobs after the initial job state)

However, we should make sure that jobs do not have an exception raised on restart (which I realize now is probably what you meant)

**Comment by garlick** [comment]
Oops, I think I misspoke.  I was thinking of whatever callback gets called for jobs that are in PRIORITY state, since jobs that were in SCHED state are bounced back to PRIORITY state on a restart.  I hadn't thought about the fact that the priority plugin might also be validating jobs at ingest, but that makes sense.

**Comment by cmoussa1** [comment]
> On a system restart, the priority plugin is loaded, then somehow all the user/bank information is updated lazily in the background (whether through rc1 script or cron etc). Therefore, currently all user submissions will be rejected until this first update happens. I think we're saying it would be better to hold all those new jobs in PRIORITY state until that first update, rather than having all the jobs be rejected.

Got it. I think that makes sense! I will open a separate issue on changing the behavior in `job.validate` to instead hold these new jobs in PRIORITY state until the first `flux account-priority-update`. Then, on an update, it can try to reprioritize all jobs with a call to `flux_jobtap_reprioritize_al ()`, which will in turn invoke the `job.priority.get` callback topic for all pending jobs, like you mentioned above.

**Comment by grondo** [comment]
Just one clarification: I do not think `job.validate` is called after a restart or when a jobtap plugin is loaded when there are active jobs. Instead (at least currently), upon reload the `job.new` callback will be run, then straight to `job.state.priority` as @garlick suggested above.

There is an open issue (flux-framework/flux-core#3812) which suggests that `job.validate` should be called instead of or in addition to `job.new` on plugin reload, to avoid plugins having to do the same thing in two different callbacks. We can move this functionality up in the queue if it would be valuable for the flux-accounting priority plugin.

Also, we should probably add some tests to flux-accounting that test restart and plugin reload scenarios, to ensure we aren't killing all the jobs in the queue in these scenarios.

**Comment by cmoussa1** [comment]
Thanks @grondo! I have an open issue also open in flux-accounting (#136) outlining the requirements for handling the plugin reloaded scenario that I haven't gotten to yet. For my own clarification, the `job.new` callback is run _instead of_ `job.validate` on active jobs when the plugin is reloaded?

I should try to make some progress there this week too, where I can add some tests to flux-accounting that test restart and plugin reload scenarios.

**Comment by grondo** [comment]
> For my own clarification, the job.new callback is run instead of job.validate on active jobs when the plugin is reloaded?

Currently that is the case, yes. The referenced issue in flux-core proposes that the `job.validate` callback be called in addition to `job.new` (though I'm not 100% sure that is the right approach at this point).

**Comment by cmoussa1** [comment]
This should be handled by #223 now, as information from the flux-accounting DB is sent to the plugin on a system restart or load with an `rc1` script, so I'll close this.
